### PR TITLE
stop all running containers when stop sand box

### PR DESCRIPTION
**Pull request by LexCC**
**What type of PR is this?**
> /kind design

**What this PR does / why we need it**:
Complete todo : https://github.com/kubernetes/kubernetes/blob/a73730147c8b45ad37a56ba210a9c721ed7a6844/pkg/kubelet/dockershim/docker_sandbox.go#L279
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kubelet: stop all running containers when stopping a Pod sand box.
```

### CODE REVIEW

File: pkg/kubelet/dockershim/docker_sandbox.go
Comment (neolit123) [inline pkg/kubelet/dockershim/docker_sandbox.go]: should this be StopContainer?

**Review by neolit123 — COMMENTED** [review]
/kind feature
/priority important-longterm

File: pkg/kubelet/dockershim/docker_sandbox.go
Comment (neolit123) [inline pkg/kubelet/dockershim/docker_sandbox.go]: is this type of async stop safe?

File: pkg/kubelet/dockershim/docker_sandbox.go
Comment (neolit123) [inline pkg/kubelet/dockershim/docker_sandbox.go]: should this be StopContainer?

**Comment by neolit123** [comment]
please add a release note instead of `None`. e.g. 
```
kubelet: stop all running containers when stopping a pod sand box.
```

File: pkg/kubelet/dockershim/docker_sandbox.go
Comment (JieJhih) [reply pkg/kubelet/dockershim/docker_sandbox.go]: Yeah.It has called docker service interface.

Comment (JieJhih) [reply pkg/kubelet/dockershim/docker_sandbox.go]: Ref. https://github.com/kubernetes/kubernetes/blob/a73730147c8b45ad37a56ba210a9c721ed7a6844/pkg/kubelet/dockershim/docker_sandbox.go#L264

File: pkg/kubelet/dockershim/docker_sandbox.go
Comment (JieJhih) [reply pkg/kubelet/dockershim/docker_sandbox.go]: I use async because of stop all containers are not ``StopPodSandbox`` main job

**Comment by neolit123** [comment]
thanks for the PR @JieJhih 
the SIG Node maintainers should have a look.
@kubernetes/sig-node-pr-reviews

**Comment by neolit123** [comment]
in the first post instead of:
> '''
> kubelet: stop all running containers when stopping a Pod sand box.
> '''

please use:
> '''release-note
> kubelet: stop all running containers when stopping a Pod sand box.
> '''

File: pkg/kubelet/dockershim/docker_sandbox.go
Comment (SataQiu) [inline pkg/kubelet/dockershim/docker_sandbox.go]: I am confused, why not `StopContainer` but `StartContainer` ?
```suggestion
			err := ds.client.StopContainer(v.ID, defaultSandboxGracePeriod)
```

Comment (SataQiu) [reply pkg/kubelet/dockershim/docker_sandbox.go]: /cc @SataQiu

Comment (JieJhih) [reply pkg/kubelet/dockershim/docker_sandbox.go]: Thanks for reminding.
Updated.

File: pkg/kubelet/dockershim/docker_sandbox.go
Comment (mattjmcnaughton) [inline pkg/kubelet/dockershim/docker_sandbox.go]:
```suggestion
```

Comment (mattjmcnaughton) [inline pkg/kubelet/dockershim/docker_sandbox.go]: Nit: If we end up merging this change, could you please delete the TODO.

**Review by mattjmcnaughton — APPROVED** [review]
LGTM modulo resolving the questions asked by @neolit123 !

For my own understanding, what mechanism did we use for cleaning up containers before this diff?

File: pkg/kubelet/dockershim/docker_sandbox.go
Comment (JieJhih) [reply pkg/kubelet/dockershim/docker_sandbox.go]: Updated
Thanks

File: pkg/kubelet/dockershim/docker_sandbox.go
Comment (feiskyer) [inline pkg/kubelet/dockershim/docker_sandbox.go]: I'm afraid this is not right. Listing all containers would also include containers in other pod sandboxes.

Comment (JieJhih) [reply pkg/kubelet/dockershim/docker_sandbox.go]: Thanks. I have updated the code , set up filter before called ``ListContainers()`` func.
I think it should work, but I have no idea why unit test failed. 
```
opts.Filters = dockerfilters.NewArgs()
f := newDockerFilter(&opts.Filters)
f.AddLabel(sandboxIDLabelKey, podSandboxID)

containers, err := ds.client.ListContainers(opts)
```

Comment (feiskyer) [reply pkg/kubelet/dockershim/docker_sandbox.go]: You can check the details for each failed cases. If you think they shouldn't happen, comment with `/retest` would rerun them.

Comment (JieJhih) [reply pkg/kubelet/dockershim/docker_sandbox.go]: Thanks. After retest all unit tests passed.

**Comment by JieJhih** [comment]
ping @SataQiu @mattjmcnaughton @feiskyer PTAL.

**Comment by SataQiu** [comment]
Thanks for your update! @JieJhih 
But please squash the commits into one, so that we can merge it.
I don't know much about this, but LGTM.

**Comment by JieJhih** [comment]
@SataQiu Squashed and pushed. Thanks

File: pkg/kubelet/dockershim/docker_sandbox.go
Comment (feiskyer) [inline pkg/kubelet/dockershim/docker_sandbox.go]: Only invoke stop container for running ones here? For already stopped containers, I don't think we need to stop again.

Comment (JieJhih) [reply pkg/kubelet/dockershim/docker_sandbox.go]: Updated. Only stop the running container.
Thanks

**Comment by xmudrii** [comment]
Hello! We have started the code freeze for 1.15. Is this PR still planned for the 1.15 cycle?

**Comment by xmudrii** [comment]
Hello! The code freeze has started last week and as this PR doesn't seem release-blocking I'm going to move it for the 1.16 release cycle.

/milestone v1.16

**Comment by ttousai** [comment]
Hello! I'm a bug triage shadow for the 1.16 release cycle and considering this issue is tagged for 1.16, but not updated for a long time, I'd like to check its status. The code freeze is starting on August 29th (about 3 days from now), which means that there should be a PR ready (and merged) by then.

Do we still target this issue to be fixed for 1.16?

**Comment by xmudrii** [comment]
As this PR has no activity since a few months ago and status is unknown, I'll remove it from the milestone.

/milestone clear

**Pull request closed without merging**
